pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_namespace_reconcile.go

```diff
@@ -0,0 +1,54 @@
+package klusterletcontroller
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	operatorapiv1 "open-cluster-management.io/api/operator/v1"
+)
+
+// namespaceReconcile is the reconciler to handle the case when spec.namespace is changed, and we need to
+// clean the old namespace. The reconciler should be run as the last reconciler, and only operate when
+// apply status is true, which ensures that old namespace should only be deleted after the agent in new
+// namespace has been deployed.
+type namespaceReconcile struct {
+	managedClusterClients *managedClusterClients
+}
+
+func (r *namespaceReconcile) reconcile(
+	ctx context.Context,
+	klusterlet *operatorapiv1.Klusterlet,
+	config klusterletConfig) (*operatorapiv1.Klusterlet, reconcileState, error) {
+	if !meta.IsStatusConditionTrue(klusterlet.Status.Conditions, operatorapiv1.ConditionKlusterletApplied) {
+		return klusterlet, reconcileContinue, nil
+	}
+	// filters namespace for klusterlet and not in hosted mo
+	namespaces, err := r.managedClusterClients.kubeClient.CoreV1().Namespaces().List(ctx, metav1.ListOptions{
+		LabelSelector: fmt.Sprintf(
+			"%s=%s",
+			klusterletNamespaceLabelKey, klusterlet.Name),
+	})
+	if err != nil {
+		return klusterlet, reconcileStop, err
+	}
+	for _, ns := range namespaces.Items {
+		if ns.Name == config.KlusterletNamespace || ns.Name == config.KlusterletNamespace+"-addon" {
+			continue
+		}
+		if err := r.managedClusterClients.kubeClient.CoreV1().Namespaces().Delete(
```

Comment:
will the old addon namespaces be deleted?

Reply:
yes

Reply:
but we should not delete old addon ns actually.